src/protocols/interaction_model/StatusCode.h

```diff
@@ -48,6 +50,108 @@ enum class Status : uint8_t
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum)
+     * @param cluster_specific_code - cluster-specific code to record with the failure
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum::kWindowNotOpen)
+     * @return a StatusCode instance properly configured.
+     */
+    template <typename T>
+    static StatusCode ClusterSpecificFailure(T cluster_specific_code)
+    {
+        static_assert(std::numeric_limits<T>::max() <= std::numeric_limits<uint8_t>::max(), "Type used must fit in uint8_t");
+        return StatusCode(Status::Failure, static_cast<uint8_t>(cluster_specific_code));
+    }
+
+    /**
+     * @brief Builder for a cluster-specific success status code.
+     *
+     * @tparam T - enum type for the cluster-specific status code
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum)
+     * @param cluster_specific_code - cluster-specific code to record with the success
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum::kBasicWindowOpen)
+     * @return a StatusCode instance properly configured.
+     */
+    template <typename T>
+    static StatusCode ClusterSpecificSuccess(T cluster_specific_code)
+    {
+        static_assert(std::numeric_limits<T>::max() <= std::numeric_limits<uint8_t>::max(), "Type used must fit in uint8_t");
+        return StatusCode(Status::Success, static_cast<uint8_t>(cluster_specific_code));
+    }
+
+    /// @return true if the core Status associated with this StatusCode is the one for success.
+    bool IsSuccess() const { return mStatus == Status::Success; }
+
+    /// @return the core Status code associated withi this StatusCode.
+    Status GetStatus() const { return mStatus; }
+
+    /// @return true if a cluster-specific code is associated with the StatusCode.
+    bool HasClusterSpecificCode() const { return mClusterSpecificCode.HasValue(); }
+
+    /// @return the cluster-specific code associated with this StatusCode or chip::NullOptional if none is associated.
+    chip::Optional<uint8_t> GetClusterSpecificCode() const
+    {
+        if (!mClusterSpecificCode.HasValue() || !((mStatus == Status::Failure) || (mStatus == Status::Success)))
+        {
+            return chip::NullOptional;
+        }
+        return mClusterSpecificCode;
+    }
+
+    // Automatic conversions to common types, using the status code alone.
+    operator Status() const { return mStatus; }
+    operator int() const { return static_cast<int>(mStatus); }
+    operator uint8_t() const { return static_cast<uint8_t>(mStatus); }
+
+private:
+    StatusCode() = delete;
+    StatusCode(Status status, uint8_t cluster_specific_code) :
+        mStatus(status), mClusterSpecificCode(chip::MakeOptional(cluster_specific_code))
+    {}
+
+    Status mStatus;
+    chip::Optional<uint8_t> mClusterSpecificCode;
```

Comment:
Perhaps we should have a static_assert that sizeof(StatusCode) <= 4?

Reply:
It doesn't per se, but if we're passing it by value it'd be really nice if it can fit in a register on 32-bit systems.  If it can't, then we make some decisions about representation vs passing by value vs just dealing with it.  But it would be good to know we are making those decisions.

Reply:
I added this.